dist/tools/compile_commands/compile_commands.py

```diff
@@ -0,0 +1,260 @@
+    else:
+        msg = "Warning: Cannot detect default include search paths for {}\n".format(compiler)
+        sys.stderr.write(msg)
+        includes = []
+    return includes
+
+
+class CompilationDetails:  # pylint: disable=too-few-public-methods
+    """
+    Representation of the compilation details stored by RIOT's build system.
+
+    :param path: Path of the module to read compilation details from
+    :type path: str
+    """
+    def __init__(self, path):
+        with open(os.path.join(path, "src_c.txt"), "r") as file:
+            self.src_c = file.read().split()
+        with open(os.path.join(path, "src_cxx.txt"), "r") as file:
+            self.src_cxx = file.read().split()
+        with open(os.path.join(path, "cflags.txt"), "r") as file:
+            self.cflags = shlex.split(file.read())
+        with open(os.path.join(path, "cxxflags.txt"), "r") as file:
+            self.cxxflags = shlex.split(file.read())
+        with open(os.path.join(path, "directory.txt"), "r") as file:
+            self.dir = file.read().strip()
+        with open(os.path.join(path, "cc.txt"), "r") as file:
+            self.cc = file.read().strip()  # pylint: disable=invalid-name
+        with open(os.path.join(path, "cxx.txt"), "r") as file:
+            self.cxx = file.read().strip()
+
+
+class State:  # pylint: disable=too-few-public-methods
+    """
+    Entity to store the current programs state
+    """
+    def __init__(self):
+        self.def_includes = dict()
+        self.flags_to_remove = set()
+        self.is_first = True
+
+
+def get_built_in_include_flags(compiler, state, args):
+    """
+    Get built-in include search directories as parameter list in JSON format.
+
+    :param compiler: Name or path of the compiler to get the include search dirs from
+    :type compiler: str
+    :param state: state of the program
+    :param args: command line arguments
+
+    :return: The -I<...> compiler flags for the built-in include search dirs in JSON format
+    :rtype: str
+    """
+    result = ""
+
+    if compiler not in state.def_includes:
+        state.def_includes[compiler] = detect_built_in_includes(compiler, args)
+
+    for include in state.def_includes[compiler]:
+        result += TEMPLATE_FLAG.format(flag='-isystem')
+        result += TEMPLATE_FLAG.format(flag=include)
+
+    return result
+
+
+def generate_module_compile_commands(path, state, args):
+    """
+    Generate section of compile_commands.json for the module in path and write it to stdout.
+
+    :param path: path of the module's bin folder to emit the compile_commands.json chunk for
+    :type path: str
+    :param state: state of the program
+    :param args: command line arguments
+    """
+    cdetails = CompilationDetails(path)
+
+    cflags = ""
+    cxxflags = ""
+    for flag in cdetails.cflags:
+        if flag not in state.flags_to_remove:
+            cflags += TEMPLATE_FLAG.format(flag=flag.replace("\\", "\\\\").replace("\"", "\\\""))
```

Comment:
Wouldn't this be much more robust if the json module were used?

Reply:
Updated to use that module. I was unaware that there is a build in module for json.

This changes the indent a bit, as I want to write out the compile command directory directly once everything for one object file is assembled. Otherwise I would have to keep everything in RAM until the very end, which might not ideal. It would be nice if there would be a 'initial_indent_level=1' parameter for `json.dumps()` for this, but I guess we can live with a slightly odd pretty printing.